bliss/models/fnp.py

```diff
@@ -0,0 +1,475 @@
+        py = self.label_vdecoder(z, u.unsqueeze(0))
+        log_py = py.log_prob(y_all).sum() / XM.size(0)
+        assert not torch.isnan(log_py)
+        obj = log_pqz + log_py
+        return obj
+
+    def calc_log_pqz(self, pz, qz, z):
+        # pylint: disable=attribute-defined-outside-init
+        """
+        Calculates the log difference between pz and qz (with an optional free bits strategy)
+        """
+        pqz_all = pz.log_prob(z) - qz.log_prob(z)
+        assert torch.isnan(pqz_all).sum() == 0
+        if self.fb_z > 0:
+            log_qpz = -torch.sum(pqz_all)
+
+            if self.training:
+                if log_qpz.item() > self.fb_z * z.size(0) * z.size(1) * (1 + 0.05):
+                    self.lambda_z = torch.clamp(
+                        self.lambda_z * (1 + 0.1), min=1e-8, max=1.0
+                    )
+                elif log_qpz.item() < self.fb_z * z.size(0) * z.size(1):
+                    self.lambda_z = torch.clamp(
+                        self.lambda_z * (1 - 0.1), min=1e-8, max=1.0
+                    )
+
+            log_pqz = self.lambda_z * pqz_all.sum()
+
+        else:
+            log_pqz = pqz_all.sum()
+        return log_pqz
+
+    def forward(self, XR, yR, XM, yM, G_in=None, A_in=None):
+        return -self.log_prob(XR, yR, XM, yM, G_in, A_in)
+
+    def predict(self, x_new, XR, yR, sample=True, A_in=None, sample_Z=True):
+        n_ref = XR.size(0)
+        u, pz = self.encode(XR, yR, x_new, None, A_in)
+        uM = u[n_ref:]
+        if sample_Z:
+            z = pz.rsample()
+        else:
+            z = pz.means
+        zM = z[:, n_ref:]
+
+        py = self.label_vdecoder(zM, uM.unsqueeze(0))
+        if sample:
+            y_pred = py.sample()
+        else:
+            y_pred = py.means
+
+        return y_pred
+
+
+## ***********************
+## FNP-specific submodules
+## ***********************
+class DepGraph(nn.Module):
+    """
+    A dependency-graph module for use within FNP.
+    For tensors of input encodings from the reference points (uR)
+    and dependent points (uM), this module returns matrices
+    that represent the dependency structure.
+
+    This implementation is based on https://github.com/AMLab-Amsterdam/FNP,
+    distributed with the MIT license. Significant changes have been made,
+    so any problems or bugs should not be attributed to the original authors.
+    """
+
+    def __init__(self, dim_u, temperature=0.3):
+        super().__init__()
+        ## Temperature for LogitRelaxedBernoulli when training
+        self.temperature = temperature
+        ## Initialized the distance function pairwise_g
+        ## This has a single learned parameter
+        self.g_logscale = nn.Parameter(torch.tensor(math.log(dim_u) * 0.5))
+        self.g = lambda x: self.logitexp(
+            -0.5
+            * torch.sum(torch.pow(x[:, dim_u:] - x[:, 0:dim_u], 2), 1, keepdim=True)
+            / self.g_logscale.exp()
+        ).view(x.size(0), 1)
```

Comment:
maybe better to define this as a separate function? Not sure it has to be a muti-line lambda(hard to parse)